beater/beater_test.go

```diff
@@ -204,43 +204,124 @@ func (d *DummyOutputClient) Close() error {
 	return nil
 }
 
-func SetupServer(b *testing.B) *http.ServeMux {
-	out := outputs.Group{
-		Clients:   []outputs.Client{&DummyOutputClient{}},
-		BatchSize: 5,
-		Retry:     0, // no retry. on error drop events
+func DummyPipeline() (*pipeline.Pipeline, error) {
+	return pipeline.New(
+		beat.Info{Name: "test-apm-server"},
+		nil,
+		func(e queue.Eventer) (queue.Queue, error) {
+			return memqueue.NewBroker(memqueue.Settings{
+				Eventer: e,
+				Events:  20,
+			}), nil
+		},
+		outputs.Group{
+			Clients:   []outputs.Client{&DummyOutputClient{}},
+			BatchSize: 5,
+			Retry:     0, // no retry. on error drop events
+		},
+		pipeline.Settings{
+			WaitClose:     0,
+			WaitCloseMode: pipeline.NoWaitOnClose,
+		},
+	)
+}
+
+func (bt *beater) client(insecure bool) (string, *http.Client) {
+	transport := &http.Transport{}
+	if insecure {
+		transport.TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
 	}
 
-	queueFactory := func(e queue.Eventer) (queue.Queue, error) {
-		return memqueue.NewBroker(memqueue.Settings{
-			Eventer: e,
-			Events:  20,
-		}), nil
+	bt.mutex.Lock() // for reading bt.server
+	defer bt.mutex.Unlock()
+	if parsed, err := url.Parse(bt.server.Addr); err == nil && parsed.Scheme == "unix" {
+		transport.DialContext = func(_ context.Context, _, _ string) (net.Conn, error) {
+			return net.Dial("unix", parsed.Path)
+		}
+		return "http://test-apm-server/", &http.Client{
+			Transport: transport,
+		}
 	}
-	testCount++
-	pip, err := pipeline.New(
-		beat.Info{Name: "testBeat"},
-		monitoring.Default.NewRegistry("testing"+string(testCount)),
-		queueFactory, out, pipeline.Settings{
-			WaitClose:     0,
-			WaitCloseMode: pipeline.NoWaitOnClose,
-		})
+	scheme := "http://"
+	if bt.config.SSL.isEnabled() {
+		scheme = "https://"
+	}
+	return scheme + bt.config.Host, &http.Client{Transport: transport}
+}
+
+func (bt *beater) wait() error {
+	wait := make(chan struct{}, 1)
 
+	go func() {
+		for {
+			bt.mutex.Lock()
+			if bt.server != nil {
+				bt.mutex.Unlock()
+				break
+			}
+			bt.mutex.Unlock()
+			time.Sleep(10 * time.Millisecond)
+		}
+		wait <- struct{}{}
+	}()
+	timeout := time.NewTimer(2 * time.Second)
+
+	select {
+	case <-wait:
+		return nil
+	case <-timeout.C:
+		return errors.New("timeout waiting server create")
+	}
+}
+
+func setupBeater(t *testing.T, ucfg *common.Config) (*beater, func()) {
+	// create a pipeline
+	pip, err := DummyPipeline()
+	assert.NoError(t, err)
+
+	// create a beat
+	apm, err := instance.NewBeat("test-apm-server", "", "")
```

Comment:
spoke with beats team about this since it felt gross - suggested it was better to duplicate the logic than pull `instance` in